### PR TITLE
Bug 2063047: Added support for full-path query log file

### PR DESCRIPTION
**Pull request by JoaoBraveCoding**
* [x] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,8 @@
 # Note: This CHANGELOG is only for the monitoring team to track all monitoring related changes. Please see OpenShift release notes for official changes.
 
+## 4.11
+- [#1587](https://github.com/openshift/cluster-monitoring-operator/pull/1589) Added support for full-path query log file
```
Comment (simonpasquier) [inline CHANGELOG.md:4]: We can omit the changelog entry because the Bugzilla doc text will take care of it.

Comment (JoaoBraveCoding) [reply CHANGELOG.md]: Done ✅

File: pkg/manifests/manifests.go [hunk 2]
```diff
@@ -1659,6 +1662,30 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, trustedCABundleCM *v1.Config
 	return p, nil
 }
 
+func (f *Factory) queryLogFileConfig(p *monv1.Prometheus, queryLogFile string) (string, error) {
+	if strings.HasPrefix(queryLogFile, "/") && !(strings.HasPrefix(queryLogFile, "/dev") || strings.HasPrefix(queryLogFile, "/prometheus")) {
+		if strings.Count(queryLogFile, "/") == 1 {
+			return "", errors.Wrap(ErrConfigValidation, "attempting to mount query log file on root \"/\"")
+		}
```
Comment (arajkumar) [inline pkg/manifests/manifests.go:1669]: nit:

Refactor a bit for readability :)

```suggestion
                dirPath := filepath.Dir(queryLogFile)
		if dirPath == "/" {
		  return "", errors.Wrap(ErrConfigValidation, `attempting to mount query log file on root "/"`)
		}
		if dirPath == "/prometheus" || dirPath = "/dev"{
		  return queryLogFile, nil
		}
		... volume mount ..
```

File: pkg/manifests/manifests.go [hunk 3]
```diff
@@ -1659,6 +1663,33 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, trustedCABundleCM *v1.Config
 	return p, nil
 }
 
+func (f *Factory) queryLogFileConfig(p *monv1.Prometheus, queryLogFile string) (string, error) {
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1666]: The function can take care of setting `p.spec.queryLogFile` directly since it already modifies the p argument.

```suggestion
func (f *Factory) setupQueryLogFile(p *monv1.Prometheus, queryLogFile string) (error) {
```

in which case the caller side would look like:

```
if err != f.setupQueryLogFile(p) {
    return nil, err
}
```

and `setupQueryLogFile()` can exit early if queryLogFile is empty.

File: pkg/manifests/manifests.go [hunk 4]
```diff
@@ -1659,6 +1663,33 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, trustedCABundleCM *v1.Config
...
+	dirPath := filepath.Dir(queryLogFile)
+	if dirPath == "/" {
+		return "", errors.Wrap(ErrConfigValidation, `attempting to mount query log file on root "/"`)
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1669]:
```suggestion
		return "", errors.Wrap(ErrConfigValidation, `query log file can't be stored on the root directory`)
```

File: pkg/manifests/manifests_test.go [hunk 5]
```diff
@@ -1194,6 +1194,74 @@ ingress:
 	}
 }
 
+func TestPrometheusQueryLogFileConfig(t *testing.T) {
+	testCases := []struct {
+		name             string
+		queryLogFilePath string
+		expected         string
+		errString        string
```
Comment (simonpasquier) [inline pkg/manifests/manifests_test.go:1202]: It's less brittle to check the error presence rather than matching the error message.

```suggestion
		errExpected        bool
```

File: pkg/manifests/manifests_test.go [hunk 6]
```diff
@@ -1194,6 +1194,74 @@ ingress:
 	}
 }
 
+func TestPrometheusQueryLogFileConfig(t *testing.T) {
```
Comment (simonpasquier) [inline pkg/manifests/manifests_test.go:1197]: the test can also check the presence/absence of the volume + volume mount.

File: pkg/manifests/manifests_test.go [hunk 7]
```diff
@@ -1194,6 +1194,74 @@ ingress:
...
+	}{
+		{
+			name:             "basic config",
+			queryLogFilePath: "/tmp/query.log",
+			expected:         "/tmp/query.log",
+			errString:        "",
+		},
+		{
+			name:             "basic config",
+			queryLogFilePath: "/dev/query.log",
```
Comment (simonpasquier) [inline pkg/manifests/manifests_test.go:1212]: I'm not sure that Prometheus would start successfully in this case :thinking:

File: pkg/manifests/manifests_test.go [hunk 8]
```diff
@@ -1194,6 +1194,74 @@ ingress:
...
+		{
+			name:             "basic config",
+			queryLogFilePath: "/query.log",
+			expected:         "",
+			errString:        "attempting to mount query log file on root",
+		},
+	}
+	for _, tt := range testCases {
+		// We need to initialize the volumeClaimTemplate otherwise we will
+		// get a SIGSEGV when creating the EmptyDir volume
```
Comment (simonpasquier) [inline pkg/manifests/manifests_test.go:1231]: can you elaborate on this issue?

File: pkg/manifests/manifests_test.go [hunk 9]
```diff
@@ -1194,6 +1194,74 @@ ingress:
...
+			errString:        "",
+		},
+		{
+			name:             "basic config",
+			queryLogFilePath: "/dev/query.log",
+			expected:         "/dev/query.log",
+			errString:        "",
+		},
+		{
+			name:             "basic config",
```
Comment (simonpasquier) [inline pkg/manifests/manifests_test.go:1217]:
```suggestion
			name:             "query log file on the TSDB storage volume",
```

File: pkg/manifests/manifests_test.go [hunk 10]
```diff
@@ -1194,6 +1194,74 @@ ingress:
...
+			queryLogFilePath: "/prometheus/query.log",
+			expected:         "/prometheus/query.log",
+			errString:        "",
+		},
+		{
+			name:             "basic config",
```
Comment (simonpasquier) [inline pkg/manifests/manifests_test.go:1223]:
```suggestion
			name:             "invalid path",
```

File: pkg/manifests/manifests_test.go (hunk 8 above)
Comment (JoaoBraveCoding) [reply pkg/manifests/manifests_test.go:1231]: Yes, so if I initialize the config with just `c := NewDefaultConfig()` then once the test run I get the error:
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xcf96f5]
```
The error happens on `manifests.go:1682` the line were we see
```
			VolumeSource: v1.VolumeSource{
				EmptyDir: p.Spec.Storage.EmptyDir,
			},
```

File: pkg/manifests/manifests.go (hunk 4 above)
Comment (JoaoBraveCoding) [reply pkg/manifests/manifests.go]: Implemented suggestion

File: pkg/manifests/manifests_test.go (hunk 6 above)
Comment (JoaoBraveCoding) [reply pkg/manifests/manifests_test.go:1198]: Added tests to volumes

File: pkg/manifests/manifests_test.go (hunk 5 above)
Comment (JoaoBraveCoding) [reply pkg/manifests/manifests_test.go]: Very true, implemented suggestion

File: pkg/manifests/manifests_test.go (hunk 7 above)
Comment (JoaoBraveCoding) [reply pkg/manifests/manifests_test.go:1213]: I'm also not sure, I'm spinning up a cluster-bot to test. If it's doesn't we can add that to the tests and allow "/dev/stdout" but forbid anything other than know outputs: /dev/null /dev/stdout

File: pkg/manifests/manifests.go [hunk 11]
```diff
@@ -1659,6 +1662,36 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, trustedCABundleCM *v1.Config
...
+	if dirPath == "/prometheus" || dirPath == "/dev" {
+		return nil
+	}
+
+	p.Spec.Volumes = append(
+		p.Spec.Volumes,
+		v1.Volume{
+			Name: "query-log-file",
+			VolumeSource: v1.VolumeSource{
+				EmptyDir: p.Spec.Storage.EmptyDir,
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1682]: we want a default emptyDir here:

```suggestion
				EmptyDir: &v1.EmptyDirVolumeSource{},
```

File: pkg/manifests/manifests_test.go (hunk 8 above)
Comment (simonpasquier) [reply pkg/manifests/manifests_test.go:1232]: ack, I've added a comment about emptyDir above that should make it possible to remove the code here.

File: pkg/manifests/manifests.go [hunk 12]
```diff
@@ -1659,6 +1662,43 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, trustedCABundleCM *v1.Config
 	return p, nil
 }
 
+func (f *Factory) setupQueryLogFile(p *monv1.Prometheus) error {
+	queryLogFile := f.config.ClusterMonitoringConfiguration.PrometheusK8sConfig.QueryLogFile
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1666]: we can handle the empty value in the function directly

```suggestion
func (f *Factory) setupQueryLogFile(p *monv1.Prometheus) error {
	queryLogFile := f.config.ClusterMonitoringConfiguration.PrometheusK8sConfig.QueryLogFile
	if queryLogFile == "" {
		return nil
	}
```

File: pkg/manifests/manifests_test.go [hunk 13]
```diff
@@ -1194,6 +1195,89 @@ ingress:
...
+			name:             "invalid file under dev",
+			queryLogFilePath: "/dev/query.log",
+			expected:         "",
+			errExpected:      true,
+		},
+	}
+	for _, tt := range testCases {
+		// We need to initialize the volumeClaimTemplate otherwise we will
+		// get a SIGSEGV when creating the EmptyDir volume
+		c, err := NewConfigFromString(`prometheusK8s:
```
Comment (simonpasquier) [inline pkg/manifests/manifests_test.go:1239]: this can be removed now?

File: pkg/manifests/manifests_test.go [hunk 14]
```diff
@@ -1194,6 +1195,89 @@ ingress:
...
+			&v1.Secret{ObjectMeta: metav1.ObjectMeta{Name: "foo"}},
+			&v1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Name: "foo"}},
+		)
+		if err != nil && !tt.errExpected {
+			t.Fatal(err)
+		}
+		if err == nil && tt.errExpected {
+			t.Fatal("Expected query log file ", tt.queryLogFilePath, " to give an error, but err is nil")
+		}
+		if !tt.errExpected {
```
Comment (simonpasquier) [inline pkg/manifests/manifests_test.go:1262]: A bit easier to read IMHO (handling error case first and return early if possible)

```suggestion
		if err != nil {
			if !tt.errExpected {
				t.Fatalf("expecting no error but got: %v", err)
			}
			return
		}
		if tt.errExpected {
			t.Fatal("Expected query log file ", tt.queryLogFilePath, " to return an error but got none")
		}
```

File: pkg/manifests/manifests_test.go [hunk 15]
```diff
@@ -1194,6 +1195,89 @@ ingress:
...
+			if p.Spec.QueryLogFile != tt.expected {
+				t.Fatal("Prometheus query log is not configured correctly")
+			}
+
+			dirPath := filepath.Dir(p.Spec.QueryLogFile)
+			if dirPath != "/prometheus" && dirPath != "/dev" {
```
Comment (simonpasquier) [inline pkg/manifests/manifests_test.go:1268]: It would be better to encode the check in the test case struct (e.g. `if tc.expectVolume { ... }`

**Comment by JoaoBraveCoding** [comment]
@simonpasquier yes indeed, I think the problem might have been fixed with my last commit https://github.com/openshift/cluster-monitoring-operator/pull/1587/commits/6983bfcc8c38885dd7d17f6f4e2386544da5fb85

**Comment by JoaoBraveCoding** [comment]
not clear to me why e3e-agnostic-operator is failing... going to simply re-test /test e2e-agnostic-operator

File: pkg/manifests/manifests.go [hunk 16]
```diff
@@ -1659,6 +1660,50 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, trustedCABundleCM *v1.Config
...
+			VolumeSource: v1.VolumeSource{
+				EmptyDir: &v1.EmptyDirVolumeSource{},
+			},
+		})
+
+	p.Spec.VolumeMounts = append(
+		p.Spec.VolumeMounts,
+		v1.VolumeMount{
+			Name:      "query-log-file",
+			MountPath: queryLogFile,
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1702]:
```suggestion
			MountPath: dirPath,
```

File: pkg/manifests/manifests.go [hunk 17]
```diff
@@ -1659,6 +1660,50 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, trustedCABundleCM *v1.Config
...
+		if base != "stdout" && base != "stderr" && base != "null" {
+			return errors.Wrap(ErrConfigValidation, `query log file can't be stored on a new file on the dev directory`)
+		}
+		return nil
+	}
+
+	p.Spec.Volumes = append(
+		p.Spec.Volumes,
+		v1.Volume{
+			Name: "query-log-file",
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1692]:
```suggestion
			Name: "query-log",
```

File: pkg/manifests/manifests.go [hunk 18]
```diff
@@ -1659,6 +1660,50 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, trustedCABundleCM *v1.Config
...
+			VolumeSource: v1.VolumeSource{
+				EmptyDir: &v1.EmptyDirVolumeSource{},
+			},
+		})
+
+	p.Spec.VolumeMounts = append(
+		p.Spec.VolumeMounts,
+		v1.VolumeMount{
+			Name:      "query-log-file",
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1701]:
```suggestion
			Name:      "query-log",
```

File: pkg/manifests/manifests.go [hunk 19]
```diff
@@ -1659,6 +1660,50 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, trustedCABundleCM *v1.Config
...
+	if queryLogFile == "" {
+		return nil
+	}
+	dirPath := filepath.Dir(queryLogFile)
+	if dirPath == "/" {
+		return errors.Wrap(ErrConfigValidation, `query log file can't be stored on the root directory`)
+	}
+
+	// /prometheus it is not necesssary to mount a volume since this is
+	// the location where prometheus will store TSDB
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1673]:
```suggestion
	// /prometheus is where Prometheus will store the TSDB so it is 
	// already mounted inside the pod (either from a persistent volume claim or from an empty dir).
```

File: pkg/manifests/manifests.go [hunk 20]
```diff
@@ -1659,6 +1660,50 @@ func (f *Factory) PrometheusK8s(grpcTLS *v1.Secret, trustedCABundleCM *v1.Config
...
+	p.Spec.QueryLogFile = queryLogFile
+	if dirPath == "/prometheus" {
+		return nil
+	}
+
+	// /dev it is not necesssary to mount a volume if the user configured
+	// the query log file to be one of the preexisting linux output streams
```
Comment (simonpasquier) [inline pkg/manifests/manifests.go:1680]:
```suggestion
	// It is not necesssary to mount a volume if the user configured
	// the query log file to be one of the preexisting linux output streams.
```

**Review by simonpasquier — COMMENTED** [review]
minor comments only, you can wait for the CI to pass before fixing them :)

File: pkg/manifests/manifests_test.go [hunk 21]
```diff
@@ -1194,6 +1194,85 @@ ingress:
...
+		},
+		{
+			name:             "query log file on the TSDB storage volume",
+			queryLogFilePath: "/prometheus/query.log",
+			expected:         "/prometheus/query.log",
+			errExpected:      false,
+			volumeExpected:   false,
+		},
+		{
+			name:             "query log file on the TSDB storage volume",
```
Comment (simonpasquier) [inline pkg/manifests/manifests_test.go:1220]:
```suggestion
			name:             "log to stdout",
```
